App viz interaction

dmfalke opened the pull request:
closes #1237 

This PR introduces the new `VisualiationPlugin` interface, and the ability for a `ComputationPlugin` to pass options to a `VisualizationPlugin`. This is needed to be able to reuse visualization with different computation plugins.

A `VisualizationPlugin` defines its own options and how to augment its behavior based on them. A `ComputationPlugin` can then create an "optioned" version of a `VisualizationPlugin` by calling the `withOptions` method.

For example, the alphadiv computation plugin looks like this:
```typescript
export const plugin: ComputationPlugin = {
  configurationComponent: AlphaDivConfiguration,
  configurationDescriptionComponent: AlphaDivConfigDescriptionComponent,
  visualizationTypes: {
    boxplot: boxplotVisualization.withOptions({
      getYAxisVariable(config) {
        if (AlphaDivConfig.is(config)) {
          return {
            entityId: config.collectionVariable.entityId,
            variableId: 'alphadiv',
          };
        }
      },
      getYAxisLabel(config) {
        if (AlphaDivConfig.is(config)) {
          return 'Alphadiv';
        }
      },
    }),
    scatterplot: scatterplotVisualization,
  },
  createDefaultComputationSpec: createDefaultComputationSpec,
};
```

You can see how it has defined `getYAxisVariable` and `getYAxisLabel`. The boxplot visualization can detect the presence of these options and act accordingly.

asizemore commented on src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:
```diff
@@ -460,19 +462,14 @@ function ScatterplotViz(props: VisualizationProps) {
-              overlayVariable?.vocabulary,
-              overlayVariable
-            );
+        response.scatterplot.config.computedVariableMetadata?.collectionVariable?.collectionVariableDetails?.map(
```
i wonder if we should have a check here to see if the overlay var is a provided var? Otherwise anytime we return a collection var as part of `computedVariableMetadata` we'll modify the overlayVocabulary. 

To be fair, i can't think of an instance where we'd return a collection var as part of the response to scatter and not use it on the overlay, but adding the check that the overlay var is provided may make the logic more human-digestible?

asizemore reviewed (commented):
Everything is so clean now!

dmfalke commented on src/lib/core/components/visualizations/VisualizationPlugin.tsx:
```diff
@@ -0,0 +1,30 @@
+}
+
+export function createVisualizationPlugin<Options>(
+  config: Omit<VisualizationPlugin<Options>, 'withOptions' | 'optoins'>
```
Thanks for catching. Fixed!

dmfalke commented on src/lib/core/components/visualizations/implementations/BoxplotVisualization.tsx:
```diff
@@ -103,19 +101,27 @@ const modalPlotContainerStyles = {
+  getYAxisLabel?: (computeConfig: unknown) => string | undefined;
+  getPlotSubtitle?: (computeConfig: unknown) => string | undefined;
+  disableShowMissingness?: boolean;
+}
```
I'm still thinking about the best way to approach this.

dmfalke commented on src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:
```diff
@@ -201,8 +195,17 @@ export const ScatterplotConfig = t.partial({
 });
 
-function ScatterplotViz(props: VisualizationProps) {
+interface Options {
```
I think we can factor these interfaces into reusable pieces and combine them together for specific viz-level options. I can make a first pass at this.

dmfalke commented on src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:
```diff
@@ -201,8 +195,17 @@ export const ScatterplotConfig = t.partial({
+  getOverlayVariable?(config: unknown): VariableDescriptor | undefined;
+  getPlotSubtitle?(config: unknown): string | undefined;
+  hideShowMissingnessToggle?: boolean;
+}
```
Sure!

dmfalke replied:
Just so I'm clear, would this control the visibility of this section?

![image](https://user-images.githubusercontent.com/365139/179300806-268d9b7d-cb01-4537-b946-9ae13f9fe7d8.png)

asizemore replied:
oh good question. Right. Yeah i think that whole section would just go away since Raw is the only Plot Mode possible

dmfalke commented on src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:
```diff
@@ -354,12 +363,10 @@ function ScatterplotViz(props: VisualizationProps) {
-    'xAxisVariable',
-    entities
-  );
+  const outputEntityId =
```
I took a short cut here. I didn't see a straightforward way to integrate the existing hook with provided variables. I will revisit it soon.

dmfalke replied:
See https://github.com/VEuPathDB/web-eda/pull/1244/commits/ce8ffc4a9c2a6b4ee3c9119f5f804c755e0b2d71 for how I addressed this.

dmfalke commented on src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:
```diff
@@ -416,12 +423,7 @@ function ScatterplotViz(props: VisualizationProps) {
-              (computation.descriptor.configuration as any).collectionVariable
-                .entityId
-            : outputEntity.id,
+          outputEntityId: outputEntity.id,
```
Yeah, it can, but we're using `outputEntity` elsewhere, so I'm not sure if provides that much benefit.

dmfalke commented:
@asizemore I think I've addressed all of your feedback, except for the following. I will continue to work on those when I can.

* Selector component override
* Factored `Options` interface
* Disabling trendlines

asizemore commented:
Sounds great @dmfalke ! Thanks so much!

asizemore commented on src/lib/core/components/visualizations/VisualizationPlugin.tsx:
```diff
@@ -0,0 +1,53 @@
+  /** Factory function to create a new plugin instance with options applied */
+  withOptions: (options: Options) => VisualizationPlugin<Options>;
+  /** Factory function to create a new plugin instance with a new selector icon */
+  withSelectorIcon: (icon: string) => VisualizationPlugin<Options>;
```
Successfully used to get the new selector icon to show up! However once i click on the icon i end up in a big type error in `InputVariables.tsx` somehow. Maybe you've seen this before?

dmfalke replied:
I will take a look!

asizemore replied:
as long as it works for you that's okay with me - i may have been implementing it incorrectly!

dmfalke replied:
fixed with https://github.com/VEuPathDB/web-eda/pull/1244/commits/49f428a4e16fe64b779afc98b5d332a5d490bac3

asizemore reviewed (commented):
half done! Confirmed the functionality looks good from the user's perspective. Will get more into the code tomorrow!

asizemore reviewed (approved):
Looks great to me!

jtlong3rd reviewed (approved):
Nicely done!

dmfalke commented on src/lib/core/components/computations/plugins/abundance.tsx:
```diff
@@ -24,11 +24,52 @@ export const AbundanceConfig = t.type({
+        }
+      },
+      hideShowMissingnessToggle: true,
+      hideTrendlines: true,
```
Ah, will update!

dmfalke commented on src/lib/core/components/computations/plugins/abundance.tsx:
```diff
@@ -24,11 +24,52 @@ export const AbundanceConfig = t.type({
+          };
+        }
+      },
+      getOverlayVariable(config) {
```
That was the (half-baked) idea. We could just as easily do something link this:

```typescript
interface ComputedVariableDetails {
  type: 'computed';
  entityId: string;
  placeholderDisplayName: string;
}

interface CollectionVariableDetails {
  type: 'collection';
  entityId: string;
  variableId: string;
}

interface XAxisVariableOptions {
  getXAxisVariable(config: unknown): ComputedVariableDetails | CollectionVariableDetails
}
```

This may be better and provide more flexibility in the future.

asizemore replied:
oh yikes sorry about the weird italics above - glad you could parse what i meant!

If it's easy, i agree your suggestion adds flexibility and i'd say it also adds a lot of clarity, so i'm a big fan!

dmfalke replied:
I think we should just stick with what we have for now. If we find that we need to support either computed or collection vars for a particular axis, we can revisit. Does that work for you @asizemore?

asizemore replied:
works for me!

dmfalke commented on src/lib/core/components/visualizations/VisualizationsContainer.tsx:
```diff
@@ -327,15 +327,19 @@ function NewVisualizationPicker(props: Props) {
-                      <vizType.selectorComponent {...vizOverview} />
+                    {vizPlugin ? (
+                      <img
+                        alt="Histogram"
```
Yes!

dmfalke replied:
fixed with https://github.com/VEuPathDB/web-eda/pull/1244/commits/0e6b43cc68b5588c335b2f8b82ef0b572d4eead3

dmfalke commented:
I'm going to do a minor version update for this. The changes are internal, but they are substantial, which probably warrants more than a patch.
